Lecture5 c

SehedaOleh opened the pull request:
UART echo works. There are a few things I would like to test more and deeper  but I need couple yours inputs. Thx.

SehedaOleh committed:
Init commit for UART.

SehedaOleh committed:
Add datasheets and initial changes in USART project.

SehedaOleh committed:
USART init. First try.

SehedaOleh committed:
Initial test done. Write trough UART. write test faild. on PC showed only </NULL>.

SehedaOleh committed:
Added driver for CP2102.

SehedaOleh committed:
Added UART terminal.

SehedaOleh committed:
Write to PC test done. All works properly.

SehedaOleh committed:
Added interrupt to receive one byte. One byte receiving done.

SehedaOleh committed:
Simple echo UART done. verification started. Method 2 in backup.

SehedaOleh committed:
Echo works.

SehedaOleh committed:
Changed Deinit function. Read and write function in separate.

rikorsev reviewed (changes requested):
Hi Oleg,
It looks good for a start. But a lot of work ahead. I would like that you change the following things:
1) Remove all unnecessary comments. It makes code more readable;
2) Put all specific to UART functions in specific module;
3) Don't use interrupts for reading. It is really unnecessary in this task.
4) Your current UART API is not really universally. I would like to have the following functions for reading and writing:
     - void uart_write(uint8_t *data, size_t len);
     - void uart_read(uint8_t *data, size_t len);
4) I didn't wind **open** and **close** functions in your code. 
If you don't know how to implement something please contact me via slack. I'm usually checking it once a day.

SehedaOleh committed:
Cleaning done. Added UART.h library with UART functions.

SehedaOleh committed:
Read function didn't work. Debugging needed. In debug receive *** error 65: access violation at 0x40021000 : no 'read' permission.

SehedaOleh committed:
Added file *.ini for debug. On simulation debug can enter to main function and on UART1_write(pString, 12); recieved
*** error 65: access violation at 0x4800000C : no 'read' permission
*** error 65: access violation at 0x4800000C : no 'write' permission

on ST-link debugger received *** error 76, line 1: command not supported
Cannot access Memory
Error while accessing a target resource. Resource perhaps not available or a wrong access was attempted.
*** error 57: illegal address (0x08000000)
Include "D:\\Dokuments\\GL\\Task\\MC_STM\\Lecture5_c\\DebugConfig\\KEIL_STM.ini"
MAP 0x40000000,0x47FFFFFF read write
*** error 76, line 1: command not supported

SehedaOleh committed:
Modify UART deinit, open and close function.

SehedaOleh committed:
UART_read function didn't work properly. debugging.

SehedaOleh committed:
Another try to complete read function.

rikorsev commented:
Hi! I've revived your code. Here is several mistakes.

1) **NEVER** put code implementation in header file as you did with **UART.h**.  Only **.c** files is a good place for code. Header files are necessary to store there only function declarations. So put all your uart implementation in **uart.c** file and put public functions declarations in uart.h file.
2) in UART1_read: 
        while ((USART1->ISR & USART_ISR_RXNE) != 0)
        {

        }
That means that program will be in the loop while RXNE is not equal to 0.
But we want that application waits until we get something in RDR register and then continue execution.
For this purpose correct condition in while loop is:
while ((USART1->ISR & USART_ISR_RXNE) **==** 0)
{

}

So, we wait until some data appears in the RDR.

3) You call read function like this:
UART1_read(pString, 10);

You pas there pointer **pString** that points on "TESTstring any types\r\n". Problem here that this string fiscally located in FLASH memory which is constant. So when you will try to write data that has been read from UART in this memory region application goes in HardFault because it is forbidden to read in Flash directly.

You need to allocate some buffer in RAM to perform read and write operations. Like that:
static uint8_t UART_Buff[10] = {0};
And then call it:
UART1_read(UART_Buff, sizeof(UART_Buff));

4) You don't need to use delay at all in your application. Loop in main function could be like that: 
  while(1)
  {
	UART1_read(UART_Buff, 10);
	UART1_write(UART_Buff, 10);
  }

But if you are need some delay, please, not use such brutal delays in your code:
void delay(int test_delay)
{
	int i = test_delay * 10000;
	while(i--);
}
use instead some hardware timer for that. You can find the way how to do that in my example:
https://github.com/rikorsev/GL_EmbeddedProCamp/blob/master/Lecture_1/Exercise_1/BlinkExample/Application/main.c

Best regards,
Oleksandr

SehedaOleh committed:
All your UART implementation puted to usart.c file and public functions declarations in usart.h file

SehedaOleh committed:
Add all changes according to Alexander comment. something go wrong. Debugging.

SehedaOleh committed:
Added delay_ms function according to Alexander recommendation.

SehedaOleh committed:
Read function works. But when I send more than one symbol troth it receive only first symbol.

rikorsev reviewed (approved):
Hi!
I have tried it and it works correct if amount of read and write bytes set as 1.
I think it is enough for this task, so clean it up and merge.
